File: code-samples/parquet-samples/api_allrecords_parquet.py
# Data retrieval example using Python 3 and parquet output, an ideal way of downloading large
#   files as it is column based and thus allows for per column data compression.

# define URL for the Disaster Declarations Summaries endpoint
baseUrl = "https://www.fema.gov/api/open/v2/DisasterDeclarationsSummaries.parquet"
import requests
import pandas as pd

# location for where the query results are saved
saveLocation = 'python_output.parquet'

def saveLargeQuery(baseUrl, saveLocation):
    try:
        # issue the api call as a stream
        with requests.get(baseUrl, stream=True) as response:
            # raises an exception when there is a server or network error
            response.raise_for_status() 

            # creates a file with if none exists, otherwise overwrites it
            #   we use the 'wb' mode to specify that we are writing as a binary so windows 
            #   does not add unneeded carriage returns
            with open(saveLocation, 'wb') as f:
                print('Saving file')
                for chunk in response.iter_content(): 
                    # we iterate through the chunks and write them to the, you can apply 
                    #   additional processing as well
                    f.write(chunk)
                print('File finished')
    except:
        # any additional logic to do when a request fails, such as trying again goes here
        print(f'file could not be downloaded, server returned ${response.status_code}')    

# Now we can use pandas to verify the record count or further analyze the data
def verifyFileDownload(saveLocation):
    print('Counting Records')
    df = pd.read_parquet(saveLocation)  # We create a dataframe with the file we just saved
    count = df[df.columns[0]].count()   # Here we are counting the number of non null records in the id column
    print('Record Count:', count)

# Run the save method
saveLargeQuery(baseUrl, saveLocation)
verifyFileDownload(saveLocation)

File: code-samples/parquet-samples/node/api_allrecords_parquet.js
/* Data retrieval example using Node.jsand the fetch library to handle http requests
 * The results of the https requests are saved to a parquet file called 'DisasterDeclaration.parquet'.
 * This is a streaming data example, ideal for large datasets
 */

import { createWriteStream } from "fs";
import parquet from "@dsnp/parquetjs";

// define our variables
const url ="https://www.fema.gov/api/open/v2/DisasterDeclarationsSummaries.parquet"; 
const fileLocation = "DisasterDeclaration.parquet";

// creating our write stream object
const writeStream = createWriteStream(fileLocation); 

// We create an async method to handle the downloading and writing of the parquet file
async function downloadFile(url, fileLocation) {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(
        `File could not be downloaded, server returned ${response.status}`
      );
    }
    console.log("saving file");
    for await (const chunk of response.body) {
      writeStream.write(chunk);
    }
    writeStream.end();
  } catch (error) {
    throw new Error(error);
  }
}

// Now that we've downloaded the file we want to manipulate it
async function testFile(fileLocation) {
  let reader = await parquet.ParquetReader.openFile(fileLocation); // an object to allow us to open parquet files
  let cursor = reader.getCursor(); // The cursor selects what we want to manipulate. It can be set to specific columns
  let record = null; // this will store the current record being manipulated
  let recordCount = 0;

  console.log("Counting records");

  // This will iterate through all the records, aditional data analysis can be performed here
  while ((record = await cursor.next())) {
    recordCount++; // In this example we are merely counting how many records we have
  }
  console.log("Record Count:", recordCount);
  await reader.close(); // After finishing the analysis, we must make sure to close the reader
}

// we initialize our program, downloading the file then running the tests after the program verifies the file is closed
downloadFile(url, fileLocation).then(() => {
  writeStream.on("finish", () => {
    console.log("file downloaded, running tests");
    testFile(fileLocation);
  });
});
